executor/src/witgen/jit/witgen_inference.rs

```diff
@@ -0,0 +1,368 @@
+use std::collections::{HashMap, HashSet};
+
+use itertools::Itertools;
+use powdr_ast::analyzed::{
+    AlgebraicBinaryOperation, AlgebraicBinaryOperator, AlgebraicExpression as Expression,
+    AlgebraicReference, AlgebraicUnaryOperation, AlgebraicUnaryOperator, Identity, LookupIdentity,
+    PhantomLookupIdentity, PolyID, PolynomialIdentity, PolynomialType, SelectedExpressions,
+};
+use powdr_number::FieldElement;
+
+use crate::witgen::{
+    global_constraints::RangeConstraintSet, jit::affine_symbolic_expression::LookupArgument,
+};
+
+use super::{
+    super::{range_constraints::RangeConstraint, FixedData},
+    affine_symbolic_expression::{AffineSymbolicExpression, Effect, ProcessResult},
+    cell::Cell,
+};
+
+/// This component can generate code that solves identities.
+/// It needs a driver that tells it which identities to process on which rows.
+pub struct WitgenInference<'a, T: FieldElement> {
+    fixed_data: &'a FixedData<'a, T>,
+    range_constraints: HashMap<Cell, RangeConstraint<T>>,
+    known_cells: HashSet<Cell>,
+    code: Vec<Effect<T, Cell>>,
+}
+
+impl<'a, T: FieldElement> WitgenInference<'a, T> {
+    pub fn new(
+        fixed_data: &'a FixedData<'a, T>,
+        known_cells: impl IntoIterator<Item = Cell>,
+    ) -> Self {
+        Self {
+            fixed_data,
+            range_constraints: Default::default(),
+            known_cells: known_cells.into_iter().collect(),
+            code: Default::default(),
+        }
+    }
+
+    pub fn known_cells(&self) -> &HashSet<Cell> {
+        &self.known_cells
+    }
+
+    pub fn code(self) -> Vec<Effect<T, Cell>> {
+        self.code
+    }
+
+    /// Process an identity on a certain row.
+    /// Returns true if this identity/row pair was fully processed and
+    /// should not be considered again.
+    pub fn process_identity(&mut self, id: &Identity<T>, row_offset: i32) -> bool {
+        let result = match id {
+            Identity::Polynomial(PolynomialIdentity { expression, .. }) => {
+                self.process_polynomial_identity(expression, row_offset)
+            }
+            Identity::Lookup(LookupIdentity {
+                id,
+                source: _,
+                left,
+                right,
+            })
+            | Identity::PhantomLookup(PhantomLookupIdentity {
+                id,
+                source: _,
+                left,
+                right,
+                multiplicity: _,
+            }) => {
+                // TODO multiplicity?
```

Comment:
That would have to be handled in the caller, not the callee, right?

Reply:
Yes, I removed it.